File: jasmin/tools/ipmatch.py
"""IP whitelist matching using the stdlib `ipaddress` module.

A *whitelist* is a string — a comma-separated list of IPv4/IPv6 addresses
and/or CIDR networks. Whitespace around entries is tolerated. Examples::

    '0.0.0.0/0'                       -> any IPv4 (default "allow all")
    '::/0'                            -> any IPv6
    '10.0.0.0/8'                      -> one /8 network
    '10.0.0.0/8, 192.168.1.5'         -> network + single host
    '10.0.0.0/8, 2001:db8::/32'       -> mixed v4+v6

An empty or `None` whitelist is treated as "nothing allowed".

The parser is tolerant and safe: any malformed entry is ignored (not
raised). This is deliberate — a bad config row should not silently crash
an authentication path. The caller can use :func:`validate_whitelist` at
config-set time to surface errors to the operator.
"""

from __future__ import annotations

import ipaddress
from typing import Iterable


def _split(whitelist: str | None) -> list[str]:
    if not whitelist:
        return []
    return [p.strip() for p in str(whitelist).split(',') if p.strip()]


def _parse_networks(whitelist: str | None) -> list:
    """Return a list of `IPv4Network` / `IPv6Network` parsed from the
    whitelist string. Malformed entries are skipped silently.
    """
    out = []
    for entry in _split(whitelist):
        try:
            out.append(ipaddress.ip_network(entry, strict=False))
        except (ValueError, TypeError):
            continue
    return out


def validate_whitelist(whitelist: str | None) -> tuple[bool, str | None]:
    """Validate a whitelist string. Returns ``(ok, error_message_or_None)``.

    An empty string / None is treated as an error (set `0.0.0.0/0` explicitly
    to allow everything). This helps the operator avoid accidentally locking
    themselves out by clearing the field.
    """
    parts = _split(whitelist)
    if not parts:
        return False, 'whitelist is empty (use "0.0.0.0/0" to allow any IPv4)'
    for entry in parts:
        try:
            ipaddress.ip_network(entry, strict=False)
        except (ValueError, TypeError) as e:
            return False, 'invalid entry %r: %s' % (entry, e)
    return True, None


def is_ip_allowed(ip: str | None, whitelist: str | None) -> bool:
    """Return True if *ip* falls inside any network in *whitelist*.

    *ip* may be an IPv4 or IPv6 address as a string (e.g. what
    ``transport.getPeer().host`` gives you). An empty *ip* or empty
    *whitelist* returns False.
    """
    if not ip or not whitelist:
        return False
    try:
        addr = ipaddress.ip_address(ip)
    except (ValueError, TypeError):
        return False
    for net in _parse_networks(whitelist):
        try:
            if addr.version != net.version:
                continue
            if addr in net:
                return True
        except (ValueError, TypeError):
            continue
    return False

File: internal/core/smpps/ipmatch.go
// Package smpps implements Jasmin's SMPP server flow (customer-facing binds). This first
// unit is the per-user IP whitelist matcher — a fork feature applied at bind time after
// password authentication.
package smpps

import (
	"fmt"
	"net"
	"strings"
)

// DefaultIPWhitelist allows any IPv4 — the fork's SmppsCredential 'ip' authorization
// default. Note it is IPv4-only: an IPv6 peer is NOT admitted by this default.
const DefaultIPWhitelist = "0.0.0.0/0"

// splitWhitelist splits a comma-separated whitelist, trimming entries and dropping empties.
func splitWhitelist(whitelist string) []string {
	var out []string
	for _, part := range strings.Split(whitelist, ",") {
		if part = strings.TrimSpace(part); part != "" {
			out = append(out, part)
		}
	}
	return out
}

// parseEntry parses one whitelist entry — a CIDR ("10.0.0.0/8") or a bare address
// ("192.168.1.5" → /32, "2001:db8::1" → /128) — returning nil when malformed. CIDR host
// bits are masked off (non-strict), matching Python's ipaddress.ip_network(strict=False).
func parseEntry(entry string) *net.IPNet {
	if _, network, err := net.ParseCIDR(entry); err == nil {
		return network
	}
	if ip := net.ParseIP(entry); ip != nil {
		if v4 := ip.To4(); v4 != nil {
			return &net.IPNet{IP: v4, Mask: net.CIDRMask(32, 32)}
		}
		return &net.IPNet{IP: ip, Mask: net.CIDRMask(128, 128)}
	}
	return nil
}

// parseNetworks parses the whitelist into networks, silently skipping malformed entries —
// deliberate, so a bad config row cannot crash the bind authentication path.
func parseNetworks(whitelist string) []*net.IPNet {
	var networks []*net.IPNet
	for _, entry := range splitWhitelist(whitelist) {
		if n := parseEntry(entry); n != nil {
			networks = append(networks, n)
		}
	}
	return networks
}

// ValidateWhitelist validates a whitelist string for the operator at config-set time. An
// empty whitelist is an error (set "0.0.0.0/0" explicitly to allow any IPv4), which
// prevents an accidental lock-out; every entry must be a valid address or CIDR.
func ValidateWhitelist(whitelist string) error {
	parts := splitWhitelist(whitelist)
	if len(parts) == 0 {
		return fmt.Errorf(`whitelist is empty (use "0.0.0.0/0" to allow any IPv4)`)
	}
	for _, entry := range parts {
		if parseEntry(entry) == nil {
			return fmt.Errorf("invalid whitelist entry %q", entry)
		}
	}
	return nil
}

// IsIPAllowed reports whether ip falls inside any network in the whitelist. An empty or
// unparseable ip, or an empty whitelist, returns false. The address family must match the
// network's, so the default "0.0.0.0/0" (IPv4) does NOT admit an IPv6 peer — a preserved
// Jasmin gotcha, not a bug to "fix" silently.
func IsIPAllowed(ip, whitelist string) bool {
	if ip == "" || whitelist == "" {
		return false
	}
	addr := net.ParseIP(ip)
	if addr == nil {
		return false
	}
	addrIsV4 := addr.To4() != nil
	for _, network := range parseNetworks(whitelist) {
		if addrIsV4 != (network.IP.To4() != nil) {
			continue // address-family mismatch
		}
		if network.Contains(addr) {
			return true
		}
	}
	return false
}
